version_control.go

```diff
@@ -246,71 +265,77 @@ func (c *VersionCache) UpdateBinary(ctx context.Context, binary string) (bool, e
 			}
 		}
 
-		// TODO(APP-10012): don't leave bad checksum binary on disk + symlinked
-		if goodBytes {
-			return false, nil
-		}
-
 		// if we're here, we have a mismatched checksum, as likely the URL changed, so wipe it and recompute later
 		if isCustomURL {
 			verData.UnpackedSHA = []byte{}
 		}
 	}
 
+	// TODO(APP-10012): don't leave bad checksum binary on disk + symlinked
+	if bytesMatch {
+		return needRestart, nil
+	}
+
 	// this is a new version
 	c.logger.Infof("new version (%s) found for %s", verData.Version, binary)
-
-	if !goodBytes {
-		c.logger.Warnw("mismatched checksum, redownloading",
-			"expected", hex.EncodeToString(verData.UnpackedSHA), "actual", hex.EncodeToString(shasum),
-			"url", verData.URL)
-		// download and record the sha of the download itself
-		verData.DlPath, err = utils.DownloadFile(ctx, verData.URL, c.logger)
-		if err != nil {
-			if isCustomURL {
-				data.brokenTarget = true
-			}
-			return needRestart, errw.Wrapf(err, "downloading %s", binary)
-		}
-
-		actualSha, err := utils.GetFileSum(verData.DlPath)
-		if err != nil {
-			return needRestart, errw.Wrap(err, "getting file shasum")
+	c.logger.Warnw("mismatched checksum, redownloading",
+		"expected", hex.EncodeToString(verData.UnpackedSHA), "actual", hex.EncodeToString(shasum),
+		"url", verData.URL, "shaErr", shaErr)
+
+	// download and record the sha of the download itself
+	verData.DlPath, urlErr = utils.DownloadFile(ctx, verData.URL, c.logger)
+	if urlErr != nil {
+		if isCustomURL && urlObj.Scheme != "file" {
+			// If the binary was downloaded from the internet cache that link as bad forever. We
+			// don't want to redownload every five seconds to learn the binary still does not
+			// work. Even if the binary backing the URL eventually changes, unfortunately.
+			//
+			// However, if the URL represents a file on disk, we should keep retrying. Consider
+			// the case where the file was overwritten in place. Being this update logic is run
+			// every five seconds, it's common to catch a binary while its being
+			// compiled/written to.
+			data.brokenTarget = true
 		}
+		return needRestart, errw.Wrapf(urlErr, "downloading %s", binary)
+	}
 
-		// TODO handle compressed formats, for now, the raw download is the same file
-		verData.UnpackedPath = verData.DlPath
-		verData.DlSHA = actualSha
+	actualSha, err := utils.GetFileSum(verData.DlPath)
```

Comment:
Everything else here should be exactly the same except for the level of indentation.